src/python/WMCore/DataStructs/Run.py

```diff
@@ -1,120 +1,197 @@
+                self.eventsPerLumi[lumi] = events
+            else:
+                self.eventsPerLumi[lumi] += events
         return self
+
     def __iter__(self):
-        return self.lumis.__iter__()
+        return self.eventsPerLumi.__iter__()
 
     def __next__(self):
-        return self.lumis.__next__()
+        """
+        __next__ no longer needed
+        """
+        raise NotImplementedError
+
     def __len__(self):
-        return self.lumis.__len__()
-    def __getitem__(self,key):
-        return self.lumis.__getitem__(key)
-    def __setitem__(self,key,value):
-        return self.lumis.__setitem__(key,value)
-    def __delitem__(self,key):
-        return self.lumis.__delitem__(key)
+        """
+        Number of lumis
+        """
+        return self.eventsPerLumi.__len__()
+
+    def __getitem__(self, key):
+        """
+        Get the nth lumi from the list (no event count)
+        """
+        return sorted(self.eventsPerLumi.keys()).__getitem__(key)
+
+    def __setitem__(self, key, lumi):
+        """
+        Replace the nth lumi from the list (no event count)
+        """
+        try:
+            oldLumi = sorted(self.eventsPerLumi.keys())[key]  # Extract the lumi from the sorted list
+            del self.eventsPerLumi[oldLumi]  # Delete it and add the new one
+        except IndexError:
+            pass
+        self.appendLumi(lumi)
+
+    def __delitem__(self, key):
+        try:
+            oldLumi = sorted(self.eventsPerLumi.keys())[key]  # Extract the lumi from the sorted list
+            del self.eventsPerLumi[oldLumi]  # Delete it
+        except IndexError:
+            pass
 
     def __eq__(self, rhs):
-        if not isinstance(rhs, Run) :
+        """
+        Check equality of run numbers and then underlying lumi/event dicts
+        """
+        if not isinstance(rhs, Run):
             return False
         if self.run != rhs.run:
             return False
-        return list(self.lumis) == list(rhs.lumis)
+        return self.eventsPerLumi == rhs.eventsPerLumi
 
     def __ne__(self, rhs):
         return not self.__eq__(rhs)
 
     def __hash__(self):
-
+        """
+        Calculate the value of the hash
+        """
         value = self.run.__hash__()
-        self.lumis.sort()
-        for lumi in self.lumis:
-            value += lumi.__hash__()
+        value += hash(frozenset(self.eventsPerLumi.items()))  # Hash that represents the dictionary
         return value
 
+    @property
+    def lumis(self):
+        """
+        Property that makes existing uses of myRun.lumis function by returning a list
+        """
+        return sorted(self.eventsPerLumi.keys())
+
+    @lumis.setter
+    def lumis(self, lumiList):
+        """
+        Setter to allow for replacement of the lumis with a list or list of tuples
+        """
+        self.eventsPerLumi = {}  # Remove existing dictionary
+        for lumi in lumiList:
+            if isinstance(lumi, (list, tuple)):
```

Comment:
Yeah. Lots of existing code uses just lists to create a run, I think with code like `myRun.lumis = [1,2,4,6]` Using this interface lets us intercept that use case and treat it as if they'd written `myRun.lumis = [(1,100), (2,150), ... ]`